### PR TITLE
doge#helpers#trim (vim version compatibility)

### PR DESCRIPTION
**Pull request by mg979**
# Prelude

Thank you for helping out DoGe!

By contributing to DoGe you agree to the following statements **(Replace `[ ]` with `[x]` with those you agree with)**:
- [ ] I have read and understood the [Contribution Guidelines](https://github.com/kkoomen/vim-doge/blob/master/CONTRIBUTING.md).
- [ ] I have read and understood the [Code of Conduct](https://github.com/kkoomen/vim-doge/blob/master/CODE_OF_CONDUCT.md).

# Why this PR?

Compatibility with older vim version.
How older? I don't know, but it works with vim 8 debian stretch.

### CODE REVIEW

**Comment by kkoomen** [comment]
What vim version are you running?

**Comment by mg979** [comment]
I'm on debian stretch at the moment, so 8.0, but without the 1630 patch that has the trim() function.

**Comment by kkoomen** [comment]
@mg979 That's still no answer to my question. What Vim version are you using? I specifically need the patch number. Open Vim and check the startup screen.

**Comment by mg979** [comment]
```
VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Jun 21 2019 04:10:35)                                                                                                                                     
Patch incluse: 1-197, 322, 377-378, 550, 649, 651, 703, 706-707                                                                                                                                        
Patch aggiuntive: 8.1.1401, 8.1.1382, 8.1.1368, 8.1.1367, 8.1.1366, 8.1.1365, 8.1.1046, 8.1.0613, 8.1.0547, 8.1.0546, 8.1.0544, 8.1.0540, 8.1.0539, 8.1.0538, 8.1.0506, 8.1.0208, 8.1.0206, 8.1.0205, 8
.1.0189, 8.1.0177, 8.1.0067, 8.1.0066                                                                                                                                                                  
Modificato da pkg-vim-maintainers@lists.alioth.debian.org                                                                                                                                              
Compilato da pkg-vim-maintainers@lists.alioth.debian.org                                                                                                                                               
Versione gigante con GUI GTK2.  Funzionalità incluse (+) o escluse (-):                                                                                                                                
+acl             +cmdline_compl   +digraphs        +folding         +langmap         +mouseshape      -mzscheme        +python3         +syntax          +toolbar         +writebackup                 
+arabic          +cmdline_hist    +dnd             -footer          +libcall         +mouse_dec       +netbeans_intg   +quickfix        +tag_binary      +user_commands   +X11                         
+autocmd         +cmdline_info    -ebcdic          +fork()          +linebreak       +mouse_gpm       +num64           +reltime         +tag_old_static  +vertsplit       -xfontset                    
+balloon_eval    +comments        +emacs_tags      +gettext         +lispindent      -mouse_jsbterm   +packages        +rightleft       -tag_any_white   +virtualedit     +xim                         
+browse          +conceal         +eval            -hangul_input    +listcmds        +mouse_netterm   +path_extra      +ruby            +tcl             +visual          +xpm                         
++builtin_terms  +cryptv          +ex_extra        +iconv           +localmap        +mouse_sgr       +perl            +scrollbind      +termguicolors   +visualextra     +xsmp_interact               
+byte_offset     +cscope          +extra_search    +insert_expand   +lua             -mouse_sysmouse  +persistent_undo +signs           +terminfo        +viminfo         +xterm_clipboard             
+channel         +cursorbind      +farsi           +job             +menu            +mouse_urxvt     +postscript      +smartindent     +termresponse    +vreplace        -xterm_save                  
+cindent         +cursorshape     +file_in_path    +jumplist        +mksession       +mouse_xterm     +printer         +startuptime     +textobjects     +wildignore                                   
+clientserver    +dialog_con_gui  +find_in_path    +keymap          +modify_fname    +multi_byte      +profile         +statusline      +timers          +wildmenu                                     
+clipboard       +diff            +float           +lambda          +mouse           +multi_lang      -python          -sun_workshop    +title           +windows                                      
   file vimrc di sistema: "$VIM/vimrc"                                                                                                                                                                 
       file vimrc utente: "$HOME/.vimrc"                                                                                                                                                               
    II file vimrc utente: "~/.vim/vimrc"                                                                                                                                                               
        file exrc utente: "$HOME/.exrc"                                                                                                                                                                
  file gvimrc di sistema: "$VIM/gvimrc"                                                                                                                                                                
      file gvimrc utente: "$HOME/.gvimrc"                                                                                                                                                              
   II file gvimrc utente: "~/.vim/gvimrc"                                                                                                                                                              
       defaults file: "$VIMRUNTIME/defaults.vim"                                                                                                                                                       
    file menu di sistema: "$VIMRUNTIME/menu.vim"                                                                                                                                                       
         $VIM di riserva: "/usr/share/vim"                                                                                                                                                             
Compilazione: gcc -c -I. -Iproto -DHAVE_CONFIG_H -DFEAT_GUI_GTK  -pthread -I/usr/include/gtk-2.0 -I/usr/lib/x86_64-linux-gnu/gtk-2.0/include -I/usr/include/gio-unix-2.0/ -I/usr/include/cairo -I/usr/i
nclude/pango-1.0 -I/usr/include/atk-1.0 -I/usr/include/cairo -I/usr/include/pixman-1 -I/usr/include/libpng16 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/libpng16 -I/usr/include/pango-1.0 -I/usr/incl
ude/harfbuzz -I/usr/include/pango-1.0 -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/freetype2 -Wdate-time  -g -O2 -fdebug-prefix-map=/build/vim-xBMBkh/vim-8.0.01
97=. -fstack-protector-strong -Wformat -Werror=format-security -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1                                                                                                   
Link: gcc   -L. -Wl,-z,relro -Wl,-z,now -fstack-protector -rdynamic -Wl,-export-dynamic -Wl,-E  -Wl,-z,relro -Wl,-z,now -Wl,--as-needed -o vim   -lgtk-x11-2.0 -lgdk-x11-2.0 -lpangocairo-1.0 -latk-1.0
 -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lpangoft2-1.0 -lpango-1.0 -lgobject-2.0 -lglib-2.0 -lfontconfig -lfreetype -lSM -lICE -lXpm -lXt -lX11 -lXdmcp -lSM -lICE  -lm -ltinfo -lnsl  -lselinux  -lacl -la
ttr -lgpm -ldl  -L/usr/lib -llua5.2 -Wl,-E  -fstack-protector-strong -L/usr/local/lib  -L/usr/lib/x86_64-linux-gnu/perl/5.24/CORE -lperl -ldl -lm -lpthread -lcrypt  -L/usr/lib/python3.5/config-3.5m-x
86_64-linux-gnu -lpython3.5m -lpthread -ldl -lutil -lm -L/usr/lib/x86_64-linux-gnu -ltcl8.6 -ldl -lz -lpthread -lieee -lm -lruby-2.3 -lpthread -lgmp -ldl -lcrypt -lm
```

**Comment by kkoomen** [comment]
What specific Vim version is your Vim mentioning in your startup screen when you run `vim -Nu NONE`? Is it `8.0.707`?

**Comment by mg979** [comment]
Yes, 8.0.707

**Comment by kkoomen** [comment]
I will look into it later. Thanks for the effort of the PR. Also, I did merge master in your branch, since I noticed that I didn't include Vim 8.0 in the travis tests, which is highly critical for this PR.

**Comment by mg979** [comment]
Ok thanks. There's another PR that is more involving (and more important I think). One test is failing though, could be because of the last commit.

**Comment by kkoomen** [comment]
I suggest you pull from the master branch and push that to your second pull request its branch.

**Comment by kkoomen** [comment]
Can you figureout which Vim version (including patch) is required now besides 8.0.1630? Otherwise I will test this.

**Comment by mg979** [comment]
I don't know how I could test it, but at least saying >800 would be safe I think.

**Comment by kkoomen** [comment]
I use the docker to built different images and then I test hem individually. I'll do that later today, thanks anyway.

File: autoload/doge/helpers.vim [hunk 1]
```diff
@@ -66,5 +66,18 @@ function! doge#helpers#placeholder(...) abort
...
 endfunction
 
+""
+" @public
+" Helper for compatibility with vim versions without the trim() function
+
+let s:no_trim = ( has('nvim') && !has('nvim-0.3.2') ) ||
+      \         ( !has('nvim') && (v:version < 800 || (v:version == 800 && !has('patch1630'))) )
+
+let s:trim_pattern = '[ \t\n\r\x0B\xA0]*\(.\{-}\)[ \t\n\r\x0B\xA0]*$'
```
[the remark this answers is not in the document]
Comment (mg979) [reply autoload/doge/helpers.vim]: The stackoverflow page is:

https://stackoverflow.com/questions/4478891/is-there-a-vimscript-equivalent-for-rubys-strip-strip-leading-and-trailing-s

But instead of just `\s*`, I used `[ \t\n\r\x0B\xA0]*` as pattern, based on `:help trim()`, so that it looks for either:
- a space
- a tab
- a new line
- a carriage return (?)
- not sure about `\x0B` (it's in help)
- `\xA0` (non-breaking space)

File: plugin/doge.vim [hunk 2]
```diff
@@ -1,15 +1,6 @@
 let s:save_cpo = &cpoptions
 set cpoptions&vim
 
-let s:unsupported_vim = !has('nvim') && (v:version < 800 || (v:version == 800 && !has('patch1630')))
```
[the remark this answers is not in the document]
Comment (mg979) [reply plugin/doge.vim]: You mean `s:unsupported_vim` ? Yes you're right, I just didn't know what the right version would be.

[the remark this answers is not in the document]
Comment (kkoomen) [reply plugin/doge.vim]: I just pushed the new minimum versions we can do. Thanks to this PR we can support 7.4. See [this](https://github.com/kkoomen/vim-doge/pull/42/commits/1a5236099ee1ed898a18239730e68f8c027bf6fb) commit for more info. 

Can you re-add the version checks with the correct versions? Thanks.

[the remark this answers is not in the document]
Comment (mg979) [reply plugin/doge.vim]: What about neovim? They're all supported now?

[the remark this answers is not in the document]
Comment (kkoomen) [reply plugin/doge.vim]: NeoVim was already supported, but instead of 0.3.2 we can also support 0.2.0 now. Check the commit I just mentioned. It shows everything we can support if this PR gets merged.

File: autoload/doge/helpers.vim (hunk 1 above)
[the remark this answers is not in the document]
Comment (kkoomen) [reply autoload/doge/helpers.vim]: I think for now the amount is okay. You're missing a caret at the beginning of the regex. 

This: 

`[ \t\n\r\x0B\xA0]*\(.\{-}\)[ \t\n\r\x0B\xA0]*$` 

should be 

`^[ \t\n\r\x0B\xA0]*\(.\{-}\)[ \t\n\r\x0B\xA0]*$`.

[the remark this answers is not in the document]
Comment (kkoomen) [reply autoload/doge/helpers.vim]: Also, `\x0b` should be vertical tab I believe.

[the remark this answers is not in the document]
Comment (mg979) [reply autoload/doge/helpers.vim]: I think `\n` and `\r` could be removed, especially if I put a caret, how could they ever appear in the middle of a line? I just put them because they are mentioned in the vim help, but I don't have a real use case for them. I tested the thing on a python script.

File: plugin/doge.vim [hunk 3]
```diff
@@ -1,11 +1,20 @@
 let s:save_cpo = &cpoptions
 set cpoptions&vim
 
-let s:unsupported_vim = !has('nvim') && (v:version < 800 && !has('patch-7.4.2219'))
+let s:unsupported_vim = !has('nvim') && (v:version < 700 && !has('patch-7.4.2219'))
```
Comment (mg979) [inline plugin/doge.vim]: What was wrong about this check? I thought vim could be version 700 but not have this patch.

I tested it in vim 7.4.160 and both `has('patch-7.4.2219')` and `has('patch2219')` return 0, the second one should be wrong anyway because it returns 0 also in later vims, the first one returns 1 also in neovim.

File: autoload/doge/helpers.vim (hunk 1 above)
[the remark this answers is not in the document]
Comment (kkoomen) [reply autoload/doge/helpers.vim]: Sure, let's shorten the regex.

File: plugin/doge.vim (hunk 3 above)
Comment (kkoomen) [reply plugin/doge.vim]: `has('patch-7.4.2219')` will return 1 in Vim 8+ as well. The `has('patch2219')` will check in your current major version if it contains the 2219 patch. If I am on Vim8 it will check for 8.x.2219 and that's not something we want. If you are below 7 or you do not have 7.4.2219 then we can throw an error. 

For NeoVim we use `has('nvim')` as an additional check.

So I will adjust the statement. It should be good after I replace the `&&` with `||`.

Comment (kkoomen) [reply plugin/doge.vim]: `has('patch-7.4.2119')` will return 1 in Vim 8+ as well. The `has('patch2119')` will check in your current major version if it contains the 2119 patch, if I am on Vim 7 it will check if I have 7.x.2119, if I am on Vim8 it will check for 8.x.2119 and that's not something we want. If you are below 7 OR you do not have 7.4.2119 then we can throw an error. 

For NeoVim we use `has('nvim')` as an additional check.

What I just did was update a typo in the version (it should've been `2119` instead of `2219` and I removed the `v:version` check, since this version is already checked now using `has('patch-7.4.2119')`. If you use Vim 8 this will still return `1`. If you are below this version you should get a `0` returned. Can you confirm this?

**Comment by kkoomen** [comment]
Ugh, tests are failing. I'll fix this tomorrow. Probably we have to go higher than Vim 7.4.2119. I'll check it tomorrow.

**Comment by mg979** [comment]
It could be because of the different pattern, a test is failing after that change, and it was ok before. I can try to revert that commit and see if it fixes it.

**Comment by mg979** [comment]
Tests run fine now, it could be the caret, is it ok if I force-push some patterns and see what's the reason?

File: plugin/doge.vim (hunk 3 above)
Comment (mg979) [reply plugin/doge.vim]: As of https://github.com/kkoomen/vim-doge/pull/42/commits/14e4ce5cb1f764934b53ae532e0bc4f8567727ee, it should be ok.

**Comment by mg979** [comment]
I added the caret but restored `\n\r`, tests pass now and I'm leaving it like this.

**Comment by kkoomen** [comment]
> Tests run fine now, it could be the caret, is it ok if I force-push some patterns and see what's the reason?

I rather you just use the docker inside the project. Simply run: `./run-tests`. Use `./run-tests --help` for more info.

> I added the caret but restored \n\r, tests pass now and I'm leaving it like this.

Cool. I was quite sure that 7.4.2119 was the minimum version we could support. Thanks for fixing it.

**Comment by kkoomen** [comment]
@mg979 Can you confirm the Vim breaks on older versions? Just to make sure the version checks are done right.

**Comment by mg979** [comment]
Yes I get the unsupported message on 7.4.

File: autoload/doge/helpers.vim [hunk 4]
```diff
@@ -68,15 +68,15 @@ endfunction
...
+" Helper for compatibility with vim versions without the trim() function.
 function! doge#helpers#trim(string) abort
-  return s:no_trim ? substitute(a:string, s:trim_pattern, '\1', '') : trim(a:string)
+  let s:no_trim = (has('nvim') && !has('nvim-0.3.2')) ||
+        \         (!has('nvim') && (v:version < 800 || !has('patch8.0.1630')))
+
+  let s:trim_pattern = '^[ \t\n\r\x0B\xA0]*\(.\{-}\)[ \t\n\r\x0B\xA0]*$'
+  return s:no_trim
+        \ ? substitute(a:string, s:trim_pattern, '\1', '')
+        \ : trim(a:string)
```
Comment (mg979) [inline autoload/doge/helpers.vim]: If you put those inside the function you can remove the s: scope, but anyway this is re-evaluating pattern and version at every function call, how is it 'optimized'? Just nitpicking, it shouldn't be important.

**Comment by kkoomen** [comment]
> Yes I get the unsupported message on 7.4.

Which patch version?

File: autoload/doge/helpers.vim (hunk 4 above)
Comment (kkoomen) [reply autoload/doge/helpers.vim]: Ah, my bad. Was in a hurry this morning. I'll fix it properly, since it _does_ matter indeed.

**Comment by mg979** [comment]
It's in 7.4 with no patches applied (welcome screen shows just 7.4).

File: autoload/doge/helpers.vim [hunk 5]
```diff
@@ -68,15 +68,15 @@ endfunction
...
-let s:no_trim = ( has('nvim') && !has('nvim-0.3.2') ) ||
-      \         ( !has('nvim') && (v:version < 800 || (v:version == 800 && !has('patch1630'))) )
-
-let s:trim_pattern = '^[ \t\n\r\x0B\xA0]*\(.\{-}\)[ \t\n\r\x0B\xA0]*$'
-
+" Helper for compatibility with vim versions without the trim() function.
 function! doge#helpers#trim(string) abort
-  return s:no_trim ? substitute(a:string, s:trim_pattern, '\1', '') : trim(a:string)
+  let s:no_trim = (has('nvim') && !has('nvim-0.3.2')) ||
+        \         (!has('nvim') && (v:version < 800 || !has('patch8.0.1630')))
```
Comment (mg979) [inline autoload/doge/helpers.vim]: You forgot a dash here, should be `has('patch-8.0.1630')`, not `has('patch8.0.1630')` (it returns 1 in vim 8.0 and it complains about trim() function not existing).

Comment (kkoomen) [reply autoload/doge/helpers.vim]: Ah, good catch. Just fixed it.

**Comment by kkoomen** [comment]
@mg979 Do you have anything else? I think we can merge into master.

**Comment by mg979** [comment]
I was saying, maybe reword last commit and force-push :smiley:  (`patch` not `path`), but it's not the last one anymore. For the rest it seems to be working in vim 8.0.

Now just saying (hoping I'm not sounding critic), but hyper-optimizing things can be harmful, last commit for example, vim is sensible about quote types for patterns and I couldn't be sure if printf() returns the right kind of quote, you could have used string concatenation but it's a waste of time mostly. I would just revert the last commit unless you're sure there's some real gain from it (I don't think there is).

**Comment by kkoomen** [comment]
>I couldn't be sure if printf() returns the right kind of quote

It seems that it does. Tests are doing perfect.

**Comment by kkoomen** [comment]
Mhm, I tried to `reword` the commit using rebase as described [here](https://coderwall.com/p/_thg9a/reword-a-git-commit-message) but that didn't do what I want (._.)

**Comment by kkoomen** [comment]
@mg979 Shall I merge?

**Comment by mg979** [comment]
For me it's ok, just one thing: how is it that tests didn't catch the trim() error before?

**Comment by kkoomen** [comment]
> how is it that tests didn't catch the trim() error before?

"the trim() error" as in: the error that you get in version 7?

**Comment by mg979** [comment]
No, when you forgot the dash in the patch version (fixed in https://github.com/kkoomen/vim-doge/pull/42/commits/1703e00d91e6727a35bb18b4e531f5b4eb3e9ce0), I had an error while executing doge in vim8, but tests were not failing.

**Comment by kkoomen** [comment]
Can you provide the full version number that you've tested?

**Comment by mg979** [comment]
Still the same as in

https://github.com/kkoomen/vim-doge/pull/42#issuecomment-529158692

**Comment by kkoomen** [comment]
That's a bit weird. The previous checks are now the checks for `s:no_trim`. The trim() function is supported since 8.0.1630+ and I did not have tests for _below_ this version. You used a version where it wasn't supported yet and I'm not going to test a version that is below the minimum supported version.

The docker versions all supported the trim() function so it didn't error on that, but we fixed that now since we support 7.4.

**Comment by mg979** [comment]
Understood, the important thing is that tests work as intended. For me it's ok!

**Comment by kkoomen** [comment]
This feature has been merged and released in [v1.12.0](https://github.com/kkoomen/vim-doge/tree/v1.12.0).

Feel free to submit any new issues if you experience any unwanted behavior in the future. Thanks for your contribution.
